File: backend/models/world.py
from pydantic import BaseModel
from uuid import UUID
from datetime import datetime
from typing import Any


class WorldChunk(BaseModel):
    """A chunk of the voxel world owned by a pet."""

    id: UUID
    pet_id: UUID
    chunk_x: int
    chunk_y: int
    chunk_z: int
    voxel_data: dict[str, Any]  # JSONB voxel data
    updated_at: datetime


class WorldSnapshot(BaseModel):
    """A diff snapshot of world changes."""

    id: UUID
    pet_id: UUID
    diff_data: dict[str, Any]
    created_at: datetime

File: frontend/src/types/world.ts
export interface Voxel {
  x: number;
  y: number;
  z: number;
  r: number;
  g: number;
  b: number;
  a: number;
  metadata_id?: string;
}

export interface Chunk {
  chunk_x: number;
  chunk_y: number;
  chunk_z: number;
  voxels: Voxel[];
}

export interface PetEntity {
  position: { x: number; y: number; z: number };
  voxels: Voxel[];
}

export const CHUNK_SIZE = 16;
